src/platforms/stlinkv3/platform.c

```diff
@@ -0,0 +1,259 @@
+#define SCB_CCR_IC_Msk (1UL << SCB_CCR_IC_Pos) /*!< SCB CCR: Instruction cache enable bit Mask */
+
+#define SCB_CCR_DC_Pos 16U                     /*!< SCB CCR: Cache enable bit Position */
+#define SCB_CCR_DC_Msk (1UL << SCB_CCR_DC_Pos) /*!< SCB CCR: DC Mask */
+
+#define SCB_CCSIDR_NUMSETS_Msk (0x7fffUL << SCB_CCSIDR_NUMSETS_Pos) /*!< SCB CCSIDR: NumSets Mask */
+#define SCB_CCSIDR_NUMSETS_Pos 13U                                  /*!< SCB CCSIDR: NumSets Position */
+
+#define SCB_CCSIDR_ASSOCIATIVITY_Pos 3U /*!< SCB CCSIDR: Associativity Position */
+#define SCB_CCSIDR_ASSOCIATIVITY_Msk (0x3ffUL << SCB_CCSIDR_ASSOCIATIVITY_Pos) /*!< SCB CCSIDR: Associativity Mask */
+#define CCSIDR_WAYS(x)               (((x)&SCB_CCSIDR_ASSOCIATIVITY_Msk) >> SCB_CCSIDR_ASSOCIATIVITY_Pos)
+
+#define SCB_DCISW_SET_Pos 5U                             /*!< SCB DCISW: Set Position */
+#define SCB_DCISW_SET_Msk (0x1ffUL << SCB_DCISW_SET_Pos) /*!< SCB DCISW: Set Mask */
+
+#define SCB_DCISW_WAY_Pos 30U                        /*!< SCB DCISW: Way Position */
+#define SCB_DCISW_WAY_Msk (3UL << SCB_DCISW_WAY_Pos) /*!< SCB DCISW: Way Mask */
+
+#define CCSIDR_SETS(x) (((x)&SCB_CCSIDR_NUMSETS_Msk) >> SCB_CCSIDR_NUMSETS_Pos)
+
+static void __DSB(void)
+{
+	__asm__ volatile("dsb 0xf" ::: "memory");
+}
+
+static void __ISB(void)
+{
+	__asm__ volatile("isb 0xf" ::: "memory");
+}
+
+static void SCB_EnableICache(void)
+{
+	__DSB();
+	__ISB();
+	SCB_ICIALLU = 0UL; /* invalidate I-Cache */
+	__DSB();
+	__ISB();
+	SCB_CCR |= (uint32_t)SCB_CCR_IC_Msk; /* enable I-Cache */
+	__DSB();
+	__ISB();
+}
+
+static void SCB_EnableDCache(void)
+{
+	uint32_t ccsidr;
+	uint32_t sets;
+	uint32_t ways;
+
+	SCB_CCSELR = 0U; /*(0U << 1U) | 0U;*/ /* Level 1 data cache */
+	__DSB();
+
+	ccsidr = SCB_CCSIDR;
+
+	sets = (uint32_t)(CCSIDR_SETS(ccsidr));
+	do {
+		ways = (uint32_t)(CCSIDR_WAYS(ccsidr));
+		do {
+			SCB_DCISW =
+				(((sets << SCB_DCISW_SET_Pos) & SCB_DCISW_SET_Msk) | ((ways << SCB_DCISW_WAY_Pos) & SCB_DCISW_WAY_Msk));
+#if defined(__CC_ARM)
+			__schedule_barrier();
+#endif
+		} while (ways-- != 0U);
+	} while (sets-- != 0U);
+	__DSB();
+
+	SCB_CCR |= (uint32_t)SCB_CCR_DC_Msk; /* enable D-Cache */
+
+	__DSB();
+	__ISB();
+}
+
+int platform_hwversion(void)
+{
+	return hw_version;
+}
+
+void platform_nrst_set_val(bool assert)
+{
+	gpio_set_val(SRST_PORT, SRST_PIN, !assert);
+	if (assert)
+		for (int i = 0; i < 10000; i++)
+			__asm__("nop");
+}
+
+bool platform_nrst_get_val()
+{
+	return gpio_get(SRST_PORT, SRST_PIN) == 0;
+}
+
+/* GND_DETECT is pull low with 100R. Probably some task should
+ * pull is high, test and than immediate release */
+#define GND_DETECT_PORT GPIOG
+#define GND_DETECT_PIN  GPIO5
+
+const char *platform_target_voltage(void)
+{
+	/* On the stlinkv3, the target input voltage is divided by two.
+	 * The ADC is sampling at 12 bit resolution.
+	 * Vref+ input is assumed to be 3.3 volts. */
+	static char ret[] = "0.0V";
+	uint8_t channels[] = {
```

Comment:
This can be simplified to `uint8_t channel = ADC_CHANNEL0;` and a single `&` added to the use below.